File: sText/src/main/java/com/samid/sTextView/base/SText.java
package com.samid.sTextView.base;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.os.Build;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewTreeObserver;

import androidx.core.view.ViewCompat;

import java.util.ArrayList;
import java.util.List;

public abstract class SText implements ISText {

    protected int mHeight, mWidth;
    protected CharSequence mText, mOldText;
    protected TextPaint mPaint, mOldPaint;
    protected STextView mSTextView;
    protected List<Float> gapList = new ArrayList<>();
    protected List<Float> oldGapList = new ArrayList<>();
    protected float progress; // 0 ~ 1
    protected float mTextSize;
    protected float oldStartX = 0;
    protected AnimationListener animationListener;

    public void setProgress(float progress) {
        this.progress = progress;
        mSTextView.invalidate();
    }

    @Override
    public void init(STextView hTextView, AttributeSet attrs, int defStyle) {
        mSTextView = hTextView;
        mOldText = "";
        mText = hTextView.getText();
        progress = 1;

        mPaint = new TextPaint(Paint.ANTI_ALIAS_FLAG);
        mOldPaint = new TextPaint(mPaint);

        mSTextView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    mSTextView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                } else {
                    mSTextView.getViewTreeObserver().removeGlobalOnLayoutListener(this);
                }
                mTextSize = mSTextView.getTextSize();
                mWidth = mSTextView.getWidth();
                mHeight = mSTextView.getHeight();
                oldStartX = 0;

                try {
                    int layoutDirection = ViewCompat.getLayoutDirection(mSTextView);

                    oldStartX = layoutDirection == View.LAYOUT_DIRECTION_LTR
                            ? mSTextView.getLayout().getLineLeft(0)
                            : mSTextView.getLayout().getLineRight(0);

                } catch (Exception e) {
                    e.printStackTrace();
                }

                initVariables();
            }
        });
        prepareAnimate();
    }

    private void prepareAnimate() {
        mTextSize = mSTextView.getTextSize();
        mPaint.setTextSize(mTextSize);
        mPaint.setColor(mSTextView.getCurrentTextColor());
        mPaint.setTypeface(mSTextView.getTypeface());
        gapList.clear();
        for (int i = 0; i < mText.length(); i++) {
            gapList.add(mPaint.measureText(String.valueOf(mText.charAt(i))));
        }
        mOldPaint.setTextSize(mTextSize);
        mOldPaint.setColor(mSTextView.getCurrentTextColor());
        mOldPaint.setTypeface(mSTextView.getTypeface());
        oldGapList.clear();
        for (int i = 0; i < mOldText.length(); i++) {
            oldGapList.add(mOldPaint.measureText(String.valueOf(mOldText.charAt(i))));
        }
    }

    @Override
    public void animateText(CharSequence text) {
        mSTextView.setText(text);
        mOldText = mText;
        mText = text;
        prepareAnimate();
        animatePrepare(text);
        animateStart(text);
    }

    @Override
    public void setAnimationListener(AnimationListener listener) {
        animationListener = listener;
    }

    @Override
    public void onDraw(Canvas canvas) {
        drawFrame(canvas);
    }

    protected abstract void initVariables();

    protected abstract void animateStart(CharSequence text);

    protected abstract void animatePrepare(CharSequence text);

    protected abstract void drawFrame(Canvas canvas);
}

File: sText/src/main/java/com/samid/sTextView/base/SText.kt
package com.samid.stextview.base

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import androidx.core.view.ViewCompat

abstract class SText : ISText {
    protected var mHeight = 0
    protected var mWidth = 0
    protected var mText: CharSequence = ""
    protected var mOldText: CharSequence = ""
    protected var mPaint: TextPaint? = null
    protected var mOldPaint: TextPaint? = null
    protected lateinit var mSTextView: STextView
    protected var gapList: MutableList<Float> = ArrayList()
    protected var oldGapList: MutableList<Float> = ArrayList()
    protected var progress = 0f
        private set
    protected var mTextSize = 0f
    protected var oldStartX = 0f
    protected var animationListener: AnimationListener? = null
        private set

    override fun init(sTextView: STextView, attrs: AttributeSet?, defStyle: Int) {
        mSTextView = sTextView
        mOldText = ""
        mText = sTextView.text
        progress = 1f
        mPaint = TextPaint(Paint.ANTI_ALIAS_FLAG)
        mOldPaint = TextPaint(mPaint)

        mSTextView.viewTreeObserver?.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                mSTextView.viewTreeObserver?.removeOnGlobalLayoutListener(this)

                mTextSize = mSTextView.textSize
                mWidth = mSTextView.width
                mHeight = mSTextView.height
                oldStartX = 0f

                try {
                    val layoutDirection = ViewCompat.getLayoutDirection(mSTextView)
                    oldStartX =
                        if (layoutDirection == View.LAYOUT_DIRECTION_LTR)
                            mSTextView.layout.getLineLeft(0) else mSTextView.layout.getLineRight(
                            0
                        )
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                initVariables()
            }
        })
        prepareAnimate()
    }

    fun setProgress(progress: Float) {
        this.progress = progress

        mSTextView.invalidate()
    }

    private fun prepareAnimate() {
        mTextSize = mSTextView.textSize ?: 0f
        mPaint!!.textSize = mTextSize
        mPaint!!.color = mSTextView.currentTextColor ?: Color.BLUE
        mPaint!!.typeface = mSTextView.typeface
        gapList.clear()
        for (i in mText.indices) {
            gapList.add(mPaint!!.measureText(mText[i].toString()))
        }
        mOldPaint!!.textSize = mTextSize
        mOldPaint!!.color = mSTextView.currentTextColor ?: Color.BLUE
        mOldPaint!!.typeface = mSTextView.typeface
        oldGapList.clear()

        for (i in mOldText.indices) {
            oldGapList.add(mOldPaint!!.measureText(mOldText[i].toString()))
        }
    }

    override fun animateText(text: CharSequence) {
        mSTextView.text = text
        mOldText = mText
        mText = text

        prepareAnimate()
        animatePrepare(text)
        animateStart(text)
    }

    override fun setAnimationListener(listener: AnimationListener) {
        animationListener = listener
    }

    override fun onDraw(canvas: Canvas) {
        drawFrame(canvas)
    }

    protected abstract fun initVariables()
    protected abstract fun animateStart(text: CharSequence?)
    protected abstract fun animatePrepare(text: CharSequence?)
    protected abstract fun drawFrame(canvas: Canvas?)
}